### ISSUE TITLE
from_generator feedback and questions

### ISSUE THREAD

**Issue by tillahoffmann** [tensorflow/tensorflow#13101]
@mrry, thank you for implementing the `from_generator` method in `tf.data`. I just wanted to provide some feedback and ask a few more questions.

# Interface

In addition to having `generator` be a callable that returns an iterator, would it be possible to support iterators that aren't wrapped in a callable? E.g. instead of 

```python
pool = multiprocessing.Pool()
dataset = tf.contrib.data.Dataset.from_generator(
    lambda: pool.imap(some_function, some_data), dtypes, shapes
)
```

also support

```python
pool = multiprocessing.Pool()
dataset = tf.contrib.data.Dataset.from_generator(
    pool.imap(some_function, some_data), dtypes, shapes
)
```

# Return types

`from_generator` does not seem to support unpacking numpy arrays at the moment. I don't think it's essential but would be a nice-to-have. E.g. this fails

```python
def generator():
    while True:
        yield np.zeros(2, np.float32)
        
dataset = tf.contrib.data.Dataset.from_generator(generator, (tf.float32, tf.float32))
x, y = dataset.make_one_shot_iterator().get_next()
session = tf.Session()
session.run([x, y])
```

but this runs smoothly

```python
def generator():
    while True:
        yield tuple(np.zeros(2, np.float32))
        
dataset = tf.contrib.data.Dataset.from_generator(generator, (tf.float32, tf.float32))
x, y = dataset.make_one_shot_iterator().get_next()
session = tf.Session()
session.run([x, y])
```

# Performance

I've played around with a *very* naive example using the generator API (in the hope to eventually leverage ipyparallel or some other distributed computing framework). Unfortunately, I can't achieve the same performance that I would get using `feed_dicts`. The setup is as follows

```python
import tensorflow as tf
from tensorflow.contrib import data as tfdata
import numpy as np
from time import time

num_batches = 1000
batch_size = 100

class Generator:
    def __init__(self):
        self.times = []
    
    def __iter__(self):
        while True:
            x = np.random.normal()
            y = 3 + 5 * x
            x, y = np.asarray([x, y], np.float32)
            self.times.append(time())
            yield x, y

generator_state1 = Generator()

dataset = tfdata.Dataset.from_generator(
    lambda: generator_state1, 
    (tf.float32, tf.float32),
    (tf.TensorShape([]), tf.TensorShape([]))
)
prefetched = dataset.prefetch(3 * batch_size)
batches = prefetched.batch(batch_size)
iterator = batches.make_one_shot_iterator()

x, y = iterator.get_next()

w = tf.Variable([0, 0], dtype=tf.float32)
prediction = w[0] + w[1] * x
loss = tf.losses.mean_squared_error(y, prediction)
optimizer = tf.train.AdamOptimizer(0.1)
train_op = optimizer.minimize(loss)
init_op = tf.global_variables_initializer()

session = tf.Session()
session.run(init_op)
```

Running the optimisation gives me

```python
losses = []

start = time()
for _ in range(num_batches):
    _, _loss = session.run([train_op, loss])
    losses.append(_loss)
time() - start  # about seven seconds
```

Doing the same using feed_dicts gives

```python
losses = []

generator_state2 = Generator()
iterator = iter(generator_state2)

start = time()
for _ in range(num_batches):
    _x, _y = np.transpose([next(iterator) for _ in range(batch_size)])
    _, _loss = session.run([train_op, loss], {x: _x, y: _y})
    losses.append(_loss)
time() - start  # about one second
```

It seems that the dataset created using the `from_generator` method isn't fetching from the generator fast enough:

```python
np.mean(np.diff(generator_state1.times))  # 7.1533812683949508e-05
np.mean(np.diff(generator_state2.times))  # 1.0633696558370612e-05
```

Thanks again for this, and looking forward to hearing your thoughts.

**Comment by Utumno** [comment]
My 2 cents of feedback: I have been trying to use the `Dataset.from_generator` API (thanks) to return an iterator to another dataset, constructed in the generator - but I can't really make it work without passing the session in:

```python
import tensorflow as tf

range5 = tf.contrib.data.Dataset.range(5).make_one_shot_iterator()
with tf.Session() as sess:
    print(tf.__version__)
    def _dataset_generator():
        while True:
            try:
                yield sess.run(range5.get_next())
            except tf.errors.OutOfRangeError:
                return
    gen_dataset = tf.contrib.data.Dataset.from_generator(_dataset_generator,
                                                         tf.int64)
    gen_it = gen_dataset.make_one_shot_iterator()
    while True:
        try:
            print(sess.run(gen_it.get_next()))
        except tf.errors.OutOfRangeError:
            break
```

I may be missing something obvious but there seems to be no way to "chain" the iterators - so just issue `yield range5.get_next()` . Moreover the pattern above does not (easily ?) generalize to more complex datasets. Implementing the generator as a python generator over a typical Dataset needs not use the session (but we still must explicitly loop in python)
And a note on errors - I erroneously wrote `Dataset.zip((Dataset.from_tensors(...), tf.range(count))` and I got the rather cryptic `errors_impl.UnknownError: AttributeError: 'Tensor' object has no attribute 'output_types'`. It seemed rather natural to pass a tensor as argument - I guess many will be tripped by this. Would help to have some type checks in zip for a clearer error message maybe?
Thanks!

**Comment by mrry** [comment]
Hi folks, thanks for the feedback. I started with a pretty conservative interface and implementation for `Dataset.from_generator()`, so we should be able to evolve it towards a better state. Responding to @tillahoffmann's and @Utumno's  individual points:

> In addition to having generator be a callable that returns an iterator, would it be possible to support iterators that aren't wrapped in a callable?

Yes, we could do this, but I wasn't able to figure out a sensible set of type checks to make this work smoothly. There are also some annoying wrinkles: e.g. a generator expression is a Python iterator so you can only iterate over it once, unless you wrap it in a callable. Of course, this only matters if you try to `.repeat()` the resulting dataset! I think the workaround (explicit wrapping) is better than the confusion that could arise here, but I'm open to better suggestions... perhaps these would involve better static information about whether it's possible to repeat a dataset?

> `from_generator` does not seem to support unpacking numpy arrays at the moment. I don't think it's essential but would be a nice-to-have.

I suppose we could extend the `nest` library so that it would delve into NumPy arrays [when it does `nest.flatten_up_to(output_types, values)`](http://google3/third_party/tensorflow/python/data/ops/dataset_ops.py?l=309&rcl=170909071) on the elements yielded from the generator. It doesn't seem incompatible with how we already use it, but I'm not sure what it would take. Let's call that "contributions welcome". :)

> I've played around with a *very* naive example using the generator API (in the hope to eventually leverage ipyparallel or some other distributed computing framework). Unfortunately, I can't achieve the same performance that I would get using `feed_dicts`.

This is a bit surprising, and we'll need to do some more investigation. It's possible that there's some overhead from acquiring and releasing the GIL in `tf.py_func()` and I don't think it's been heavily optimized. One quick question though: is the steady-state performance 7x worse, or does the difference in mean shrink if you exclude the first step when the iterator is constructed? (That's not an excuse of course, but it would help to focus our efforts!) 

> I have been trying to use the Dataset.from_generator API (thanks) to return an iterator to another dataset, constructed in the generator - but I can't really make it work without passing the session in[.]

This is not really the expected use for `Dataset.from_generator()` and as [we discussed on Stack Overflow](https://stackoverflow.com/a/46557087/3574081), you can use `Dataset.map()` if you need to capture a different iterator. However, I'm not quite sure what the use case is here... why don't you want to use the original iterator directly?

> I erroneously wrote `Dataset.zip((Dataset.from_tensors(...), tf.range(count))` and I got the rather cryptic `errors_impl.UnknownError: AttributeError: 'Tensor' object has no attribute 'output_types'`.

I coincidentally fixed this in 3c00952c6680d77ee2f10def35fbc7cbd138aea3.

Thanks again for using the new API, and let me know if you have more comments or thoughts on these answers!

**Comment by Utumno** [comment]
> This is not really the expected use for Dataset.from_generator() and as we discussed on Stack Overflow, you can use Dataset.map() if you need to capture a different iterator. However, I'm not quite sure what the use case is here... why don't you want to use the original iterator directly?

Use case: I have a dataset (d1, lengths) of variable length tensors (read from tfrecord files) and I want to produce from it another dataset d2 with _slices_ of the tensors in d1.

> I coincidentally fixed this in 3c00952.

I was bitten by the list vs tuple in Dataset.zip and took some to figure that out too - good job!

**Comment by mrry** [comment]
@Utumno Would `Dataset.flat_map()` work for this? For example, if you have a dataset of variable-length tensors called `var_len_dataset`, you can get all the slices as follows:

```python
var_len_dataset = ...  # Element shape (?, x, y, z)
all_slices = var_len_dataset.flat_map(
    lambda x: tf.data.Dataset.from_tensor_slices(x))  # Element shape (x, y, z)
```

**Comment by Utumno** [comment]
I have the zip(d1, num_slices) dataset and I want to produce another dataset that slices each tensor in d1 into num_slices slices (segments of  equal length). I posted a [SO question](https://stackoverflow.com/q/46604371/281545) - I know the code there is not correct (I should be using the dummy dataset) but still I am not sure the approach is at all possible.

**Comment by mrry** [comment]
I posted [an answer](https://stackoverflow.com/a/46608404/3574081): `Dataset.flat_map()` should work for your case.

**Comment by tillahoffmann** [comment]
@mrry, thanks for the feedback.

> There are also some annoying wrinkles: e.g. a generator expression is a Python iterator so you can only iterate over it once, unless you wrap it in a callable.

Yes, very good point.

Regarding the performance issues: https://gist.github.com/tillahoffmann/53f902212c1c511e6c3d45abb5f37b9b

**Comment by tillahoffmann** [comment]
@mrry, do you have any more thoughts on the speed differences with `feed_dict`s, queues and the dataset API?

**Comment by mrry** [comment]
I haven't had time to dig into this in depth, but I did notice that when I dial the batch size down to 1, the `Dataset.from_generator()` version is faster than using a `feed_dict`. This suggests that the overhead might be coming from the `Dataset.batch()` transformation. I'm surprised that batching a `std::vector<Tensor>` (even by a pretty naive approach) would be slower than stacking together a batch of ndarrays (which implicitly happens when you feed the list), but perhaps there's something to be learned from the NumPy implementation?

/cc @jsimsa who has looked at batch performance in other contexts.

**Comment by jsimsa** [comment]
@tillahoffmann give [map_and_batch](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/data/python/ops/batching.py), providing identity for the map a try. `Dataset.batch()` is serializing memory copies during batch creation, which can affect performance for large batches. `Dataset.map_and_batch()` is doing the memory copies in parallel.

**Comment by tillahoffmann** [comment]
@mrry, @jsimsa, thank you for the updates.

I have changed the line
```python
batches = dataset.batch(BATCH_SIZE).prefetch(PREFETCH_FACTOR * BATCH_SIZE)
```
to
```python
batches = dataset.apply(map_and_batch(lambda a, b: (a, b), BATCH_SIZE)).prefetch(PREFETCH_FACTOR * BATCH_SIZE)
```
in the [notebook](https://gist.github.com/tillahoffmann/53f902212c1c511e6c3d45abb5f37b9b) linked [above](https://github.com/tensorflow/tensorflow/issues/13101#issuecomment-334788891). But the code hangs indefinitely when I try to train.

**Comment by jsimsa** [comment]
@tillahoffmann apply the following fix to your version of TF: https://github.com/jsimsa/tensorflow/pull/1 and try again ... if that does not help, please use gdb to connect to the running process using `sudo gdb -p <PID>`, then run `thread apply all bt` to collect all stack traces, and share them here

**Comment by tillahoffmann** [comment]
@jsimsa, thank you for the fix. It seems to have made it into master because I can run the cod using the nightly build. Unfortunately, the performance is slightly worse than the original code that used `batch` instead of `map_and_batch`. I have updated the notebook.

**Comment by jsimsa** [comment]
@tillahoffmann I digged into this a bit more and my conclusion is that the slowdown is due to the (constant) overhead of `tf.data` runtime. In particular, I  have done the following:

```
import time
import numpy as np
import tensorflow as tf

from tqdm import tqdm
from tensorflow.contrib import data as tfdata

BATCH_SIZE = 128
FEATURE_SHAPE = (16, 16)
NUM_BATCHES = 1024*2
PREFETCH_FACTOR = 4

weights = np.random.normal(0, 1, FEATURE_SHAPE)

def data_generator():
  for _ in range(NUM_BATCHES * BATCH_SIZE):
    feature = np.random.normal(0, 1, FEATURE_SHAPE)
    value = np.sum(feature * weights) + np.random.normal(0, 1)
    yield feature.astype(np.float32), value.astype(np.float32)


def data_generator_batch():
  for _ in range(NUM_BATCHES):
    features, values = np.empty((BATCH_SIZE, 16, 16)), np.empty((BATCH_SIZE))
    for i in range(BATCH_SIZE):
      features[i] = np.random.normal(0, 1, FEATURE_SHAPE)
      values[i] = np.sum(features[i] * weights) + np.random.normal(0, 1)
    yield features, values


def build_graph(batchX, batchY):
  weights_tensor = tf.get_variable('weights', FEATURE_SHAPE, tf.float32)
  predictions = tf.reduce_sum(weights_tensor * batchX, axis=(1, 2))
  loss_tensor = tf.losses.mean_squared_error(batchY, predictions)

  optimizer = tf.train.AdamOptimizer()
  train_op = optimizer.minimize(loss_tensor)
  return loss_tensor, train_op


def train(sess, train_op, loss_tensor):
  times = []
  losses = []
  for i in tqdm(range(NUM_BATCHES)):
    start = time.time()
    _, loss = sess.run([train_op, loss_tensor])
    times.append(time.time() - start)
    losses.append(loss)
  return np.asarray(times), np.asarray(losses)

times = {}
losses = {}

# Dataset without batch training
print("Dataset with batching generator")
with tf.Graph().as_default():
  dataset = tfdata.Dataset.from_generator(
      data_generator_batch, (tf.float32, tf.float32), ((BATCH_SIZE, 16, 16),
                                                       (BATCH_SIZE)))
  batches = dataset.prefetch(PREFETCH_FACTOR)
  iterator = batches.make_one_shot_iterator()
  batch_X, batch_y = iterator.get_next()

  loss_tensor, train_op = build_graph(batch_X, batch_y)

  sess = tf.Session()
  sess.run(tf.global_variables_initializer())
times['datasets_batch'], losses['datasets_batch'] = train(
    sess, train_op, loss_tensor)

# Using generator
print("Feed with batching generator")
with tf.Graph().as_default():
  batch_X = tf.placeholder(tf.float32, (BATCH_SIZE, ) + FEATURE_SHAPE)
  batch_y = tf.placeholder(tf.float32, (BATCH_SIZE, ))

  loss_tensor, train_op = build_graph(batch_X, batch_y)

  sess = tf.Session()
  sess.run(tf.global_variables_initializer())
  tf.train.start_queue_runners(sess)

data = data_generator_batch()
_times = []
_losses = []

for i in tqdm(range(NUM_BATCHES)):
  X, y = next(data)
  start = time.time()
  _, loss = sess.run([train_op, loss_tensor], {batch_X: X, batch_y: y})
  _times.append(time.time() - start)
  _losses.append(loss)
times['generator_batch'] = np.asarray(_times)
losses['generator_batch'] = np.asarray(_losses)

# Dataset with batch training
print("Dataset without batching generator")
with tf.Graph().as_default():
  dataset = tfdata.Dataset.from_generator(
      data_generator, (tf.float32, tf.float32), (FEATURE_SHAPE, ()))
  batches = dataset.batch(BATCH_SIZE).prefetch(PREFETCH_FACTOR)
  iterator = batches.make_one_shot_iterator()
  batch_X, batch_y = iterator.get_next()

  loss_tensor, train_op = build_graph(batch_X, batch_y)

  sess = tf.Session()
  sess.run(tf.global_variables_initializer())
times['datasets_no_batch'], losses['datasets_no_batch'] = train(
    sess, train_op, loss_tensor)

# Using generator
print("Feed without batching generator")
with tf.Graph().as_default():
  batch_X = tf.placeholder(tf.float32, (BATCH_SIZE, ) + FEATURE_SHAPE)
  batch_y = tf.placeholder(tf.float32, (BATCH_SIZE, ))

  loss_tensor, train_op = build_graph(batch_X, batch_y)

  sess = tf.Session()
  sess.run(tf.global_variables_initializer())
  tf.train.start_queue_runners(sess)

data = data_generator()
_times = []
_losses = []

for i in tqdm(range(NUM_BATCHES)):
  X = []
  y = []

  for i in range(BATCH_SIZE):
      _X, _y = next(data)
      X.append(_X)
      y.append(_y)

  start = time.time()
  _, loss = sess.run([train_op, loss_tensor], {batch_X: X, batch_y: y})
  _times.append(time.time() - start)
  _losses.append(loss)
times['generator_no_batch'] = np.asarray(_times)
losses['generator_no_batch'] = np.asarray(_losses)
```

which results in:

```
Dataset with batching generator
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2048/2048 [00:09<00:00, 214.56it/s]
Feed with batching generator
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2048/2048 [00:07<00:00, 291.40it/s]
Dataset without batching generator
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2048/2048 [01:05<00:00, 31.11it/s]
Feed without batching generator
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2048/2048 [00:08<00:00, 227.84it/s]
```

In particular, if a generator that produces whole batches is used (as opposed to a generator that produces single elements batched by `tf.data`), the throughput of `tf.data`-based execution is comparable to the throughput of feed-based execution.

Normally, I would expect the cost of fetching a single element from a pipeline to be include a possible expensive `tf.data.map` operation, over which the constant overhead of executing the `tf.data` a pipeline (which for instance involves scheduling TF ops) would be amortized. For pipelines that do very little computation (such as the one used in your examples), the overhead of the general purpose pipeline execution engine can be non-negligible (as illustrated by your experiment) and pushing the batching logic into the generator is one possible solution to avoid the slowdown.

**Comment by bhack** [comment]
@jsimsa Somebody is suggesting [this approach](https://stackoverflow.com/questions/47086599/parallelising-tf-data-dataset-from-generator) on stackoverflow. Is it a workaround or a best practice?

**Comment by jsimsa** [comment]
I commented on the Stack Overflow thread; if possible, moving expensive processing from a `generator` to a `map` will make it possible to parallelize the processing. I find that preferrable to creating multiple identical generators and combining them using `interleave`.

On a related note, the extreme version of the approach discussed on the Stack Overflow page is to replace  `tf.data.Dataset.from_generator` with `tf.data.Dataset.range(N).map(...)` where `map(n)` generates the `n`-th element. If possible, this will maximize the opportunity for parallelism as the only computation that needs to be serialized is the generation of the `map` function inputs.

**Comment by bhack** [comment]
@jsimsa I think that `generator` is an important parallel case to handle in the docs cause frequently it needs to handle itself the concept of epoch.

**Comment by boeddeker** [comment]
@jsima Using `tf.data.Dataset.range(N).map(...)` instead of `tf.data.Dataset.from_generator` does not alwais work. As I mentioned in #14448, `tf.data.Dataset.from_generator` wrapps `tf.py_func` to have a generalized `tf.py_func` (i.e. allow nested structures as output instead of only flat outputs. Also it assigns the shapes. In #15121 is a PR that extracts the generalized `tf.py_func`). 

On Stack Overflow I gave an example how you can replace `tf.data.Dataset.from_generator` with something that can be parallized from tensorflow.

**Comment by bhack** [comment]
@boeddeker So mainly  `num_parallel_calls` will be independent from `dataset.batch`: I.e. `num_parallel_calls` could be set to the number of available core and `dataset.batch` to the device capacity? It this use case is `dataset.batch %  num_parallel_calls` needed to be 0?

**Comment by boeddeker** [comment]
@bhack Afaik is `dataset.batch` independent from `num_parallel_calls`. Using the example code for `from_indexable` from Stack Overflow, below a small example. The execution time is `time to get one item` devided by `num_parallel_calls` and is independent from batch size.

```python
import time

class PyDataSet:
    def __len__(self):
        return 20
        
    def __getitem__(self, item):
        time.sleep(0.5)  # <----------------- Takes 0.5 seconds to "load" the data
        return np.random.normal(size=(10))

ds = from_indexable(PyDataSet(), output_types=tf.float64, output_shapes=[None, 10], num_parallel_calls=4)
ds = ds.batch(9)
it = ds.make_one_shot_iterator()
entry = it.get_next()
with tf.Session() as sess:
    start = time.perf_counter()
    try:
        while True:
            print(sess.run(entry).shape)
    except tf.errors.OutOfRangeError:
        pass
    delta = time.perf_counter() - start
    print(delta)  # 2.5277008840057533
```

**Comment by bhack** [comment]
Yes what I meant is that if you make 4 parallel calls (i.e. you have 4 cores) for each get_next how much get_next internally will be called to fill a batch of 9?

**Comment by boeddeker** [comment]
I added a print to `__getitem__` and it printed 12 times for 4 cores with batch size 9. So there are 12 `__getitem__` calls.

Further, I changed the sleep to only apply to one entry per `num_parallel_calls`
```
        if item in [0, 4, 8]:
            time.sleep(1)
```
and the slowest entry defined the runtime (i.e. 3 seconds, ideally it would take 1 second).
So my conclusion is that `num_parallel_calls` follows the fork and join idea. It starts `num_parallel_calls` and waits until all of them have finished before launch the next `num_parallel_calls`.

Maybe @mrry knows more about the internals, my knowledge is based on toy examples.

**Comment by bhack** [comment]
Yes what I am guessing if there is not an explicitly `data.prefetch` where `dataset.batch % num_parallel_calls` is going? Is it in the next batch? Does it make sense to maintain `dataset.batch % num_parallel_calls == 0`?

**Comment by boeddeker** [comment]
I am not sure if I understand your question correctly.

My example does not contain `data.prefetch` and `dataset.batch % num_parallel_calls` is `1`.
My experiments showed that the iterator items are independent of the value `num_parallel_calls`.
The only reason that I can find for `dataset.batch % num_parallel_calls == 0` is when you are only interested in a single call and does not want to burn/waste computation power and/or RAM.

`map` caches the not required calculations for the next batch.

Here an example what I understand that happens:
`num_parallel_calls == 4`
`batch == 9`
launch `__getitem__[i] for i in [0, 1, 2, 3]`
wait until all `__getitem__` calls finished, launch `__getitem__[i] for i in [4, 5, 6, 7] `
wait until all `__getitem__` calls finished, launch `__getitem__[i] for i in [8, 9, 10, 11] `
wait until `__getitem__[8]` finished than build a batch
Note: In the background the remaining calls to `__getitem__` may still compute)

nest session run
wait until all `__getitem__` calls finished
wait until all `__getitem__` calls finished, launch `__getitem__[i] for i in [12, 13, 14, 15]`
...

**Comment by bhack** [comment]
>The only reason that I can find for dataset.batch % num_parallel_calls == 0 is when you are only interested in a single call and does not want to burn/waste computation power and/or RAM.

Yes.
Where `13, 14, 15`  will go? In the next batch?

**Comment by boeddeker** [comment]
Yes, the next batch will contain the items `9, 10, 11, 12, 13, 14, 15, 16, 17`.

**Comment by bhack** [comment]
So there is some kind of buffer that store 9, 10, 11 from the third `__getitem__` call?

**Comment by boeddeker** [comment]
Yes, from my experiments it seems to be so.

**Comment by jsimsa** [comment]
@boeddeker your understanding is almost complete ... the only difference is that in reality there is no barrier of size `num_parallel_calls`. The is `__getitem__[i + num_parallel_calls]` launched only after `__getitem__[i]` finishes but possibly before `__getitem__[i + 1]` finishes. Internally, the `map` implementation has `num_parallel_calls` cyclic buffer that is used for storing the invocation results.

**Comment by boeddeker** [comment]
@jsimsa Thanks for the explanation. 
I had a bad example, where the element `0, num_parallel_calls, ...` are the slowest.

Do you also know, why calling `20` times `time.sleep(1)` with  `num_parallel_calls=20` takes 5 seconds on a computer with 4 cores (I expect 1 second)?
It seems that the "prefetching" rule in `map` follows your explanation, but the numbers of threads are limited to `min(num_parallel_calls, cpu_core_count)`.

**Comment by bhack** [comment]
@jsimsa So basically it has a internal auto prefecthing for the next batch. Right?

**Comment by jsimsa** [comment]
Each `map` invocation itself does not have a thread pool, instead it schedules the invocations using the `interop_thread_pool` (which has a default size of `cpu_core_count`). Try increasing the size of the `interop_thread_pool` to `20` and see what effect it has on the runtime.

**Comment by boeddeker** [comment]
Thanks, `inter_op_parallelism_threads` work. But the second `tf.Session` requires also `use_per_session_threads` to overwrite the value.
```python
config = tf.ConfigProto(
    inter_op_parallelism_threads=20,
    # use_per_session_threads=True  # required when not first Session in python interpreter
)
```

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by davidparks21** [comment]
I had opened another issue (https://github.com/tensorflow/tensorflow/issues/16343) which is also related to feedback of the generator process. I think it's beneficial to move those comments here and close that issue in order to consolidate the related discussions to one issue.

---

We're given hundreds of data files, each containing many gigabytes worth of sample data in a custom format. As far as I can tell there are only two approaches to extract samples from this using `Dataset`:

1) `tf.data.Dataset.from_generator(generator=my_custom_reader, ...)`

Create a generator which produces samples. This approach is not ideal because this method must be the first dataset in the chain. The generator cannot accept a tensor. Therefore you can't batch and shuffle your list of 100's of filenames (or anything more complex). You also can't make use of `interleave(...)` because the generator can't accept a tensor, and this use case is begging to use `interleave(...)`.

A solution here might be to provide a method for a generator to accept a tensor, as `tf.py_func(...)` does for functions.

2) `tf.data.Dataset.map(map_func=tf.py_func(my_custom_reader, ...), ...)`

The map function does allow us to shuffle and parallelize the filenames using all of the functionality of the Dataset pipeline, however, with `map`, the files must be read into memory completely, and these files are large. Reading numerous files into memory is infeasible.

A solution here might be to extend the `map` function to support generators.

Unless there's an alternative approach, which I didn't glean from the docs or stackoverflow, then this seems to be an inherent limitation and a seemingly reasonable use case on which to base a feature request.

I've implemented this process using map functions by splitting large files into filename/index pairs so as to specify a subset of the file to read in each map call. This was not a trivial task. A final sample solution can be found on stack overflow below. I continue to believe that Dataset should be extended to better support this workflow. In particular with respect to supporting generators throughout the pipeline.

https://stackoverflow.com/questions/48471688/using-tensorflows-dataset-pipeline-how-do-i-name-the-results-of-a-map-oper/48589464#48589464

Update to this post:
---------------
The solution referenced above doesn't work in the case where each sample needs to be different shapes. This code example demonstrates an attempt to return an ndarray of ndarray's which produces an `unsupported object type numpy.ndarray` error in tensorflow. The goal of the example is to demonstrate reading 2 samples of variable length from a file. 

```
import tensorflow as tf
import numpy as np

def mypymap(data_numpy):
  result = np.array([
    np.array([100,101,102]),
    np.array([200,201,202,203])
  ])
  return result

data_input = ['fileA','fileB','fileC']

ds = tf.data.Dataset.from_tensor_slices(data_input)
ds = ds.map(lambda data_tensor: tf.py_func(mypymap, [data_tensor], Tout=[tf.int64], stateful=False))

element = ds.make_one_shot_iterator().get_next()

with tf.Session() as sess:
  for _ in range(3):
    print(sess.run(element))
```
So that leaves the only solution being to incorporate a bunch of logic into my generator (for interleaving, parallel file reads, and shuffling) and putting the generator up at the beginning of the preprocessing pipeline. All things that are better done as part of the `Dataset` pipeline.

**Comment by bhack** [comment]
Can we document the relationship between `from_generators` and Dataset cache?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by LynnHo** [comment]
@mrry I saw dataset.batch(BATCH_SIZE).prefetch(FACTOR * BATCH_SIZE) and dataset.batch(BATCH_SIZE).prefetch(FACTOR), which ones is correct usage?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ychfan** [comment]
@mrry Is it possible to add `inp` argument like `tf.py_func` for `Dataset.from_generator`?
It would be more flexible for this [approach](https://stackoverflow.com/a/47874044/9644463)

**Comment by carlthome** [comment]
I'm interested in something like this too @ychfan.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by RylanSchaeffer** [comment]
Has there been any discussion on or a conclusion to adding `num_parallel_calls` to `from_generator()`?

**Comment by boeddeker** [comment]
@RylanSchaeffer the function `from_generator` will never get a `num_parallel_calls` becasuse a python generator is something that cannot be parallelized. In #14448 I suggested to add an alternative to `from_generator` that can have a `num_parallel_calls` (Use `__getitem__` instead if of `__iter__`), but the feedback was that it is not necessary.

I updated my answer in https://stackoverflow.com/a/47884927/5766934, so using that code is now simpler.

**Comment by carlthome** [comment]
Assuming I/O is the bottleneck, does it suffice to chain `from_generator` with `prefetch`?

You could always do multiple `from_generator` in a `flat_map` too. Particularly `tf.contrib.data.parallel_interleave` is sweet.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by jianxucsm** [comment]
@mrry 

Hi Derek;

seems to me that the from_generator can not catch up the requirement of the data feed for the convolution neuron network processing on gpu.

following is the image of the gpu performance

![image](https://user-images.githubusercontent.com/22060946/43689057-1b46e08c-98a9-11e8-86a6-c39e32dea312.png)

I am using the dataset.from_generator to get the input data from slicing of pandas dataframe. set the prefetech as 5. the image shows the teeth of the gpu utilization which implies the gpu is waiting for the data feed in, wonder why the from_generator can not catch up the requirement via multithreading of cpu.

pls indicate if there is any trick to improve the feed speed of the from_generator.

tks.

**Comment by jianxucsm** [comment]
@mrry 

further will from_generator introduce the parameter as num_parallel_calls which will parallel the process of the generator function.

tks.

**Comment by boeddeker** [comment]
@jianxucsm As I mentioned earlier, a python generator cannot run in parallel. So from_generator will never get a num_parallel_calls, as long the input is a generator.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 106 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @mrry: It has been 120 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by zh794390558** [comment]
@boeddeker So what is the preferred method?

**Comment by boeddeker** [comment]
@zh794390558 It depends, when you have a lightweight generator, then using tf.data.Dataset.from_generator is fine.

When your generator gets expensive, I would rewrite it to something that is indexable and use my code that I posted in https://stackoverflow.com/a/47884927/5766934.

Alternatives are splitting your generator in a lightweight generator and a map function (https://stackoverflow.com/a/47089278/5766934) or use tf.contrib.data.parallel_interleave (https://stackoverflow.com/a/47874044/5766934). I personally do not like tf.contrib.data.parallel_interleave in cases where loading the examples is independent.

**Comment by zh794390558** [comment]
https://stackoverflow.com/a/47884927/5766934 this need the dataset must be a iterator, how to using the `yield` pipeline?    
The tf.contrib.data.parallel_interleave  will working easy for `yield` pipeline, and I think the tf.contrib.data.parallel_interleave has a random case  to feed data to shuffle buffer.     

Can you give a performance comparison with tf.contrib.data.parallel_interleave, TFRecord?

**Comment by boeddeker** [comment]
In my opinion is a `yield` pipeline not necessary. The code that you write in your generator is usually something like:
```python
def gen():
  # pre calculations
  for i in range(10000):
    # example calculations
    yield (i, [1] * i)
```
where you have a `for` loop over an index or a file list.
The counterpart code where this is indexable is then:
```python
class gen:
  def __init__(self):
    # pre calculations
  def __len__(self):
    return 10000
  def __getitem__(self, i):
    # example calculations
    return (i, [1] * i)
```
So most `yield` pipelines could be rewritten with a class that has `__getitem__`.
Main advantages:
- `__getitem__` can be parallelized, while a generator can not be parallelized
- you can shuffle without actually loading the data (i.e. no buffer is needed)

Since I do not use tf.contrib.data.parallel_interleave, TFRecord, I have no performance comparison.
I like to do a shuffle across all my data for each epoch and TFRecord does not fit this use case.

**Comment by bhack** [comment]
Can we add something in the doc/faq?

**Comment by boeddeker** [comment]
IMHO would be the best place to document something like that at a `from_indexable` method in the Dataset source code.
But @mrry argued in #14448 that `from_indexable` would be to trivial and the user can write the code himself.

**Comment by zh794390558** [comment]
@boeddeker why not to using tf.data.Dataset.shuffle ？If using it, then we not need to control shuffle manually in `class gen`.

**Comment by boeddeker** [comment]
@zh794390558 Because that function uses a buffer.
Usually the buffer size is significant smaller than your number of examples, because the size of your dataset is far away from fitting in memory (you can overcome this when your actual data loading happens in tensorflow).
When this is the case, the shuffle is only a local shuffle.
So when your complete dataset is sorted (e.g., the deprecated `sklearn.datasets.fetch_mldata('MNIST original')` was sorted) the shuffle can only shuffle the similar examples that are close together.

Toy example:
Let's consider your dataset is sorted and the labels are:
`[1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4, 4, 5, 5, 5, 5, 6, 6, 6, 6, 7, 7, 7, 7, 8, 8, 8, 8, 9, 9, 9, 9]`
When the buffer size is 4 you can never observe a 3, 4, 5, 6, 7, 8 or 9 at the beginning of an epoch.

**Comment by carlthome** [comment]
>Because that function uses a buffer.

Speaking of which, @mrry mentioned on a related issue (maybe StackOverflow, don't remember sorry!) that the tf.data team is looking into a disk shuffle of the entire index. Have there been any more news/work around this? I guess there are complications with distributed training but it doesn't feel like it should be that hard to write a permutation.index file per shard and maintain close to random access in the underlying data store.

`tf.data` is great but one of the big missing features is definitely the possibility of shuffling an entire `Dataset` per epoch. After porting some `keras.utils.Sequence` code to `tf.data` I get somewhat worse final test accuracy on a bunch of standard benchmarks (FashionMNIST, Boston house pricing, etc.) and it's greatly affected by shuffle buffer size.

**Comment by zh794390558** [comment]
So I  think `tf.data` should using `iterator` which same to `class gen`, then do index shuffle by `tf.data`.

**Comment by Karel-van-de-Plassche** [comment]
In my pipeline dataset shuffling is the bottleneck. Is my use-case (large dataset, small networks. In my case 3e8 samples, 9D input space, 1D output space. Networks 2 layers of 128 neurons) any consideration? Found this way of shuffling 4x faster then TF dataset shuffle in some cases:

```
perm = np.arange(self._num_examples)
np.random.shuffle(perm)
self._data = np.take(self._data, perm, axis=0)
```

**Comment by bionicles** [comment]
how do you take TFRecords of sequence examples and turn them into ragged tensors? It seems like this might be supported but there are no docs about it.

**Comment by ArtemisZGL** [comment]
hello, if my generator yield x and y, x has shape(batch, 257, 250,1) and y has shape(batch, class_num), what value should output_type be ? i try and get `generator` yielded an element that did not match the expected structure may times

**Comment by mrry** [comment]
Thanks for all the great feedback on this thread! Since this is now an old issue and the `Dataset.from_generator()` API and implementation have evolved, I'm going to close the thread now. Please feel free to open new issues if you have comments about `Dataset.from_generator()` or how it could be made more useful.

**Comment by Anhaoxu** [comment]
@mrry @aselle @davidparks21 @tillahoffmann @jsimsa 
Hi , I'm not familiar with tf.data api and huge data training tricks，in my case, I found the IO is a problem: GPU uti is low, about 15%. I use a generator function to read local multiple pickle files(about 200K),so each batch is a local file in my case, examples are bellow:
```
def train_data():
    while True:
        for i in shuffle(train_id):
            df = pd.read_pickle(df_train_val['file_name'].iloc[i],
                                compression='gzip')
            feature = df.to_numpy().reshape(df.shape[0], -1)
            yield feature.reshape(-1, 56),feature.reshape(-1, 56)
```
Then
```
gn = tf.data.Dataset.from_generator(train_data, output_types=(tf.float32, tf.float32),
                                          output_shapes=((None, 56),(None, 56))
                                         ).prefetch(tf.data.experimental.AUTOTUNE)
```
I found prefetch not work and IO bottleneck is very high, so any advice for this case, how can I speed up the training process?

**Comment by davidparks21** [comment]
@Anhaoxu, this kind of question should be moved to stack overflow, could you repost it there and put a link to the topic here?
In brief, for high IO data Python is a major obstacle here. The Python GIL (global interpreter lock) limits you to one CPU core when processing is done in python and not external libraries like tensorflow (which handles processing in C external to the Python runtime). I run a model that chews on about 400 MB/sec of data and that requires about 3 cores to handle IO. If your main process tries to do all the work it can delay the training loop, and one core might be enough for the IO period. There are 2 solutions as I have come to understand. 1) write your data to tfrecords files in subprocesses/multiprocessing (what I currently do). Tensorflow can deserialize tfrecords files in more than one core outside of Python so you're good. But that's a little cumbersome to implement. 2) Use python 3.8 (for which tensorflow is just now starting to support) and use System V shared memory where you use subprocesses (Python multiprocessing) to do the IO and prep work, and hand off a pointer to the prepared sample to the main process. All other forms of Python serialization between processes engage the Python GIL and don't work under high IO scenarios. I have not tested this approach yet, I just installed Python 3.8 and TF 2.2RC2 to do so in the past few weeks.

**Comment by boeddeker** [comment]
> In brief, for high IO data Python is a major obstacle here. The Python GIL (global interpreter lock) limits you to one CPU core when processing is done in Python

I have to correct that statement. The Python GIL limits you to execute the python code on a single CPU, not the C code.

When a function is implemented in C, it can release the GIL. When the GIL is released, you can do parallel IO with threads in python.
A famous example for a function that releases the GIL is `open`, so an IO operation usually releases the GIL. Another example are many numpy functions. So using a threadpool for numpy operations can utilizes all cores of your system (Assuming, the arrays are large enough)

@Anhaoxu The problem in your example is, that a generator cannot be executed in parallel. When you start your dataset with a range and do the IO in a tensorflow map function, the IO can be executed in parallel. That is the reason, why I proposed to have a `from_indexable`.

**Comment by davidparks21** [comment]
@boeddeker, your clarifications make an excellent point and clarifies something I wasn't clear about. Many libraries, Tensorflow, in particular, release the GIL and can process on multiple cores. However, at high IO data rates (a good fraction of a GB/sec) the problem becomes CPU bound not IO bound because just the deserialization process inside of Python drives up the CPU utilization past 1 core. 

For example, TensorFlow utilizes 3-4 cores to deserialize the tfrecords files I pass it, and those are in a nice efficient protobuf format. Python cannot read at those rates because the GIL becomes CPU bound on the basic serialization processes that are required to prepare the data for tensorflow, not on the IO-wait part.

I've spent many iterations trying to solve this problem for a relatively small CNN model that reads raw uncompressible scientific data from a 3TB dataset. With threads, I maybe get 120% CPU use out of Python. Subprocesses can't pass data to the main process because serialization becomes CPU bound in Python (GIL issue). AsyncIO with multiple event threads is all GIL based, even using c-based `uvloop`. 

If someone is using Python/Tensorflow where you see more than about 1.2 cores being used by the main process (the max I've ever seen) I would absolutely love to know about it. Perhaps the only difference in my case is that my data has to be read over a remote filesystem, so it's coming in over HTTP from a local S3-compliant interface, perhaps that changes the equation for me.

I went around and around with this problem and the only thing that worked was dumping to tfrecords and letting tensorflow handle the multi-core deserialization outside of Python. And I believe shared memory is another solution. But trying to push a good fraction of a GB/sec of data through the same python interpreter as the training loop is hopeless in my experience (at least pre-py3.8 shared-memory).

**Comment by Anhaoxu** [comment]
@boeddeker @davidparks21 
I just want to use map api to yield parallel generators, but the map api not supported generator type, in my case, I would like to know how to realize this page [https://www.tensorflow.org/guide/data_performance#parallelizing_data_extraction](url), or how to use from_indexable to realize parallel generators, it's not clear to me, and I can't find related information.
in fact, I have multiple files, they are from s3, I just found reading from s3 is slower than read from local file system, so I compressed these files, maybe 200K files, in this process  I also can concat them to be less number, maybe just 1K files, but I think this operation not solve the problem fundamentally, the problem is how can not let the GPU wait for data?
for example:
I want to shuffle their file id to make every epoch training using all files, so I write this function:
```
def train_data():
    while True:
        for i in shuffle(train_id):
            df = pd.read_pickle(df_train_val['file_name'].iloc[i],
                                compression='gzip')
            feature = df.to_numpy().reshape(df.shape[0], -1)
            yield feature.reshape(-1, 56),feature.reshape(-1, 56)
```
this function is very clear, right? so how this function be parallel? if there are some tutorial about parallel and distributed data transformation or training, I would be very appreciate.

**Comment by 183amir** [comment]
As others have already mentioned, please ask this question on stackoverflow @Anhaoxu

**Comment by boeddeker** [comment]
> your clarifications make an excellent point and clarifies something I wasn't clear about. Many libraries, Tensorflow, in particular, release the GIL and can process on multiple cores. However, at high IO data rates (a good fraction of a GB/sec) the problem becomes CPU bound not IO bound because just the deserialization process inside of Python drives up the CPU utilization past 1 core.

@davidparks21 good point. Yes, it highly depends on the data that you use. I mainly work with wav files on a remote file system. Wav files are usually relative large ~ 5MB. 
Because of that (and the file format) I have a purely IO bottleneck.
Depending on the file format and the reader also other data can be efficiently loaded (e.g. decompress images could be done without the GIL in C).
I have some good experiences for parallel deserialization of uncompressed data with `pickle.loads` (note with an `s`).

> I've spent many iterations trying to solve this problem for a relatively small CNN model that reads raw uncompressible scientific data from a 3TB dataset. With threads, I maybe get 120% CPU use out of Python. Subprocesses can't pass data to the main process because serialization becomes CPU bound in Python (GIL issue). AsyncIO with multiple event threads is all GIL based, even using c-based uvloop.

I am not sure, how you moved the data from one subprocess to another, but probably you used pickle.
Did you try to save your data as pickle, read all data with open and then call `pickle.dumps`? This can be done in threads and need only low CPU time (pickle data is already prepared for python).

@davidparks21: When you want to continue this discussion, should we move this discussion to somewhere else?
